### DISCUSSION TITLE
4-bit KV Cache

### DISCUSSION THREAD

**Discussion by Dampfinchen** [ggml-org/llama.cpp#5932] [Ideas]
Turboderp, developer of Exllama V2 has made a breakthrough: A 4 bit KV Cache that seemingly performs on par with FP16. Here are his words:

"I'm working on some benchmarks at the moment, but they're taking a while to run. Preliminary results show the Q4 cache mode is more precise overall than FP8, and comparable to full precision. HumanEval tests are still running."

https://www.reddit.com/r/LocalLLaMA/comments/1b9571u/80k_context_possible_with_cache_4bit/

This is huge. Sadly, Llama.cpp doesn't even have full 8 bit cache right now (only K cache). So in that aspect, there's a lot of potential for improvement.

**Comment by puppetm4st3r** [comment]
thats is huge! hope you can implement in llama.cpp :)

**Comment by BarfingLemurs** [comment]
Here is some benchmarks and information - https://github.com/turboderp/exllamav2/blob/master/doc/qcache_eval.md

**Comment by Dampfinchen** [comment]
> Here is some benchmarks and information - https://github.com/turboderp/exllamav2/blob/master/doc/qcache_eval.md

Very interesting. What do @ggerganov @JohannesGaessler and @slaren think about these results? The current consensus is that 4 bit KV cache isn't worth it as the uptick in perplexity would be too severe. However, that isn't the case with Turboderp's implementation. I wonder what llama.cpp can learn from that.

**Reply by JohannesGaessler** [reply]
10 rows of 512 tokens is nowhere near enough data to make conclusive statements. More generally, a metric like KL divergence would be a better choice to judge the precision loss.

**Reply by ggerganov** [reply]
After (and if) #5021 is finished we can add support for cache quantization since the V cache will no longer be transposed

**Reply by henk717** [reply]
Any updates on this @ggerganov ?

**Reply by ggerganov** [reply]
Haven't started looking into this. When I find the time, I'd like to try to extend the Metal FA kernels to support quantized KV cache and see if this make sense to add. Though no ETA atm

**Reply by JohannesGaessler** [reply]
One of my next FlashAttention goals is to adapt the mmq/mmvq code in order to handle quantized KV cache data without having to dequantize to FP16 (because that would need a lot of memory). Speed may be somewhat suboptimal though for large batch sizes on modern GPUs since mmq does not use tensor cores.

**Reply by ggerganov** [reply]
Why do you need to adapt the MMQ kernels - you just need to add support for the FA kernels, no?

**Reply by JohannesGaessler** [reply]
The MMQ *kernels* all use the same template with quantization type specific functions for allocation, data loading, and the actual dot products. I meant the latter when I was talking about the MMQ *code*. I'm thinking the way to go is to write a new template for FA and re-use the functions.

**Reply by henk717** [reply]
I have had overwhelming feedback from users that they are eager for 4-bit cache because it helps them run larger models they currently need to resort to EXL2 for. They prefer Llamacpp based solutions but when it comes down to being able to run a model and not being able to run a model on their limited VRAM they prefer to be able to run a model.

So from a user demand perspective at least on our community demand has been high for this and its something I expect the majority of our users to turn on if implemented (and given the nature of the request I expect this to be the same in the entire ecosystem) .  I fully understand the "if it makes sense to add" is from a coding perspective and the end result of implementations can sway it one way or the other, but I did want to ensure that the interest from the users is known to help guide any decisions / evaluations if it's down to user interest.

**Reply by JohannesGaessler** [reply]
I mean, in September of last year I had already made a [PR](https://github.com/ggerganov/llama.cpp/pull/2969) for quantizing the KV cache so I definitely agree that it would be a very good feature to have. But that PR was never merged due to introducing too many changes so for a new implementation I'll have to be careful to lay sufficient groundwork in advance.

**Reply by Dampfinchen** [reply]
On that note, what happened to your enhanced mmq code with tensor core support, Johannes? I remember being pretty impressed by the speed increase and low VRAM usage back then. 

https://github.com/ggerganov/llama.cpp/pull/4801#issuecomment-1901007642

Perhaps this would be useful for large batch sizes when you are working on the 4 bit kv cache?

**Reply by JohannesGaessler** [reply]
I at some point noticed that there are numerical issues with my implementation that are not easily fixable (I just never posted about this). After talking to an NVIDIA engineer I'm also thinking that I'll need to overhaul the implementation in general. But I have definitely not forgotten about this. My next three major goals are FlashAttention, n-gram lookup, and int8 tensor core support. After that I'll look into training.

**Comment by kir-gadjello** [comment]
Integral part of good performance of turboderp's KV-cache quantization is Hadamard transform for smoothing the kv distribution: 
<img width="1091" alt="Screenshot 2024-04-30 at 08 37 16" src="https://github.com/ggerganov/llama.cpp/assets/111190790/5d148abe-0bc0-436e-8587-8518acd3aa99">
See https://arxiv.org/pdf/2404.00456 and this part of turboderp's implementation https://github.com/turboderp/exllamav2/commit/324404ebe4e3c4dd0447ffc1290c312de1df02be

**Comment by henk717** [comment]
Now that #5021 is merged i'd love to see this, we have a lot of users who are eager for 4-bit KV cache to help save vram.

**Reply by strawberrymelonpanda** [reply]
Count me as one of them, but the new Flash Attention flag _really_ helps already.  Combined with this, sounds amazing.

**Comment by WiseFarAI** [comment]
This might be useful:

1. https://github.com/ClubieDong/QAQ-KVCacheQuantization
2. https://arxiv.org/abs/2404.00456

**Comment by sammcj** [comment]
Howdy 👋, 

I'm trying to find out if anyone measured the perplexity / performance with llama.cpp's q4_0 / q8_0 K/V cache that could share their results?

I've been running it with q8_0 for the past day and the vRAM savings really are excellent - brings memory usage inline with ExllamaV2, while I haven't _noticed_ any quality degradation I've only been using newer models I'm not that familiar with so would love to see some perplexity and performance measurements if you've taken any.

Sorry if I missed this in the docs somewhere 😅 

(I've got a [PR up to Ollama](https://github.com/ollama/ollama/pull/5894) to add support for this)

**Reply by JohannesGaessler** [reply]
https://github.com/ggerganov/llama.cpp/pull/7412

**Reply by sammcj** [reply]
Oh that's perfect, thank you so much! My search foo must have failed me.

**Reply by sammcj** [reply]
@JohannesGaessler I read your comment #7527 regarding the need to set `GGML_CUDA_DISABLE_GRAPHS=1` - is that still required?

**Reply by JohannesGaessler** [reply]
No, that has been fixed.

**Reply by sammcj** [reply]
Fantastic, thank you for all your hard work!

**Comment by pomoke** [comment]
On Qwen2-7B, q4_0 produces weird results. q8_0 is ok, but I haven't compared with f16 in performance.

**Reply by ddh0** [reply]
Qwen2 uses 8x GQA if I remember correctly, I'd bet that a heavily quantized KV cache hurts it more than a model with less aggressive GQA

**Reply by ag2s20150909** [reply]
type_k set to  `q8_0` and type_v set to `q4_0 ` can get right  results.

**Comment by sammcj** [comment]
FYI for anyone looking for some more recent perplexity measurements - I recently compared Q8_0 and 
Q4_0 quantisation levels on a Qwen 2.5 Coder 7B model as the Qwen series
 of models is known to be quite sensitive to quantisation.

The results shown below show the difference in quality is minimal with an increase of 0.0043

Model | Params | Quant | K/V Cache Type | ppl | Notes | Date
-- | -- | -- | -- | -- | -- | --
qwen2.5 coder | 7b | q6_k | f16/f16 | 8.3891 +/- 0.02016 | unsloth 128k gguf | 2024-12-06
qwen2.5 coder | 7b | q6_k | q8_0/q8_0 | 8.3934 +/- 0.02017 | unsloth 128k gguf | 2024-12-06

```shell
./llama-perplexity -ngl 99999 -fa -b 2048 -c 6114 -sm none \
  #   -ctk f16 -ctv f16 \
  # and
  #   -ctk q8_0 -ctv q8_0 \
  -f wiki.train.raw.txt \
  -m Qwen2.5-Coder-7B-Instruct-128k-Q6_K.gguf
```

**Reply by dpmm99** [reply]
You said you compared Q8_0 and Q4_0 but you posted a table with f16/f16 and q8_0/q8_0?

**Reply by FNsi** [reply]
Yep, how's the q4_0?
